### PR TITLE
[amazonminitv] Add extractor

### LINKED ISSUE #3576: Amazon MiniTV (https://www.amazon.in/minitv)
### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I'm running yt-dlp version **2022.04.08** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/ytdl-org/youtube-dl#can-you-add-support-for-this-anime-video-site-or-site-which-shows-current-movies-for-free) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

India

### Example URLs

- Single episode: https://www.amazon.in/minitv/tp/74271cb6-e19e-445b-a648-e15a3912b8ef

### Description

Free OTT service by Jeff uncle. 
Not related to Amazon Prime.
100% DRM free!

### Verbose log

```shell
yt-dlp -vU https://www.amazon.in/minitv/tp/74271cb6-e19e-445b-a648-e15a3912b8ef
[debug] Command-line config: ['-vU', 'https://www.amazon.in/minitv/tp/74271cb6-e19e-445b-a648-e15a3912b8ef']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, err utf-8, pref UTF-8
[debug] yt-dlp version 2022.04.08 [7884ade65]
[debug] Python version 3.9.5 (CPython 64bit) - Linux-5.11.0-49-generic-x86_64-with-glibc2.33
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: avconv -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] Checking exe version: avprobe -bsfs
[debug] exe versions: none
[debug] Optional libraries: brotli, certifi, Cryptodome, secretstorage, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.04.08, Current version: 2022.04.08
yt-dlp is up to date (2022.04.08)
[debug] [generic] Extracting URL: https://www.amazon.in/minitv/tp/74271cb6-e19e-445b-a648-e15a3912b8ef
[generic] 74271cb6-e19e-445b-a648-e15a3912b8ef: Requesting header
WARNING: [generic] Could not send HEAD request to https://www.amazon.in/minitv/tp/74271cb6-e19e-445b-a648-e15a3912b8ef: HTTP Error 405: Method Not Allowed
[generic] 74271cb6-e19e-445b-a648-e15a3912b8ef: Downloading webpage
WARNING: [generic] Falling back on generic information extractor.
[generic] 74271cb6-e19e-445b-a648-e15a3912b8ef: Extracting information
[debug] Looking for video embeds
ERROR: Unsupported URL: https://www.amazon.in/minitv/tp/74271cb6-e19e-445b-a648-e15a3912b8ef
Traceback (most recent call last):
  File "/home/z8512/.local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1408, in wrapper
    return func(self, *args, **kwargs)
  File "/home/z8512/.local/lib/python3.9/site-packages/yt_dlp/YoutubeDL.py", line 1478, in __extract_info
    ie_result = ie.extract(url)
  File "/home/z8512/.local/lib/python3.9/site-packages/yt_dlp/extractor/common.py", line 641, in extract
    ie_result = self._real_extract(url)
  File "/home/z8512/.local/lib/python3.9/site-packages/yt_dlp/extractor/generic.py", line 4039, in _real_extract
    raise UnsupportedError(url)
yt_dlp.utils.UnsupportedError: Unsupported URL: https://www.amazon.in/minitv/tp/74271cb6-e19e-445b-a648-e15a3912b8ef
```

### PR DESCRIPTION
**Pull request by nyuszika7h**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

---

### Description of your *pull request* and other information

Implements support for Amazon Mini TV, an Indian mobile-only streaming service which is distinct from Amazon Prime Video and DRM-free.

Closes #3576.

### CODE REVIEW

File: yt_dlp/extractor/minitv.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        'url': 'minitv:amzn1.dv.gti.280d2564-584f-452f-9c98-7baf906e01ab',
+        'only_matching': True,
+    }, {
+        'url': 'minitv:280d2564-584f-452f-9c98-7baf906e01ab',
+        'only_matching': True,
+    }]
+
+    def _call_api(self, asin, data=None, note=None):
+        query = {}
+        headers = copy(self._HEADERS)
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:70]:
```suggestion
        headers = self._HEADERS.copy()
```

File: yt_dlp/extractor/minitv.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+            note='Downloading title info')
+
+        prs = self._call_api(asin, note='Downloading playback info')
+
+        formats = []
+        for type_, asset in prs['playbackAssets'].items():
+            if not isinstance(asset, dict):
+                continue
+            if type_ == 'hls':
+                formats.extend(self._extract_m3u8_formats(
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:182]: _extract_m3u8_formats_and_subtitles

File: yt_dlp/extractor/minitv.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+
+        duration = traverse_obj(title_info, ('description', 'contentLengthInSeconds'))
+        chapters = []
+        credits_time = try_get(title_info, lambda x: x['timecode']['endCreditsTime'] / 1000)
+        if credits_time is not None and duration is not None:
+            chapters.append({
+                'start_time': credits_time,
+                'end_time': duration,
+                'title': 'End Credits',
+            })
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:200]:
```suggestion
        credits_time = try_get(title_info, lambda x: x['timecode']['endCreditsTime'] / 1000)
        chapters = [{
            'start_time': credits_time,
            'end_time': duration,
            'title': 'End Credits',
        }] if credits_time and duration else []
```

File: yt_dlp/extractor/minitv.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+
+        info = {
+            'id': asin,
+            'title': title_info['name'],
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:204]:
```suggestion
            'title': title_info.get('name'),
```

File: yt_dlp/extractor/minitv.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+            'series_id': title_info.get('seriesId'),
+            'season_number': title_info.get('seasonNumber'),
+            'season_id': title_info.get('seasonId'),
+            'episode_number': title_info.get('episodeNumber'),
+        }
+        if title_info.get('vodType') == 'EPISODE':
+            info.update({
+                'episode': title_info['name'],
+                'episode_id': asin,
+            })
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:225]:
```py
is_episode = title_info.get('vodType') == 'EPISODE'
return {
    ...
    'episode': title_info.get('name') if is_episode else None,
    'episode_id': asin if is_episode else None,
})
```

File: yt_dlp/extractor/minitv.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
+import json
+from copy import copy
+
+from .common import InfoExtractor
+from ..utils import ExtractorError, int_or_none, traverse_obj, try_get
+
+
+class MiniTVIE(InfoExtractor):
+    _VALID_URL = r'(?:https?://(?:www\.)?amazon\.in/minitv/tp/|minitv:(?:amzn1\.dv\.gti\.)?)(?P<id>[a-f0-9-]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:9]: why the keyword?

File: yt_dlp/extractor/minitv.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+                'episode': title_info['name'],
+                'episode_id': asin,
+            })
+
+        return info
+
+
+class MiniTVSeasonIE(MiniTVIE):
+    IE_NAME = 'minitv:season'
+    _VALID_URL = r'minitv:season:(?:amzn1\.dv\.gti\.)?(?P<id>[a-f0-9-]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:232]: this doesnt have URL?

File: yt_dlp/extractor/minitv.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+      }
+    }
+  }
+}
+''',
+            },
+            note='Downloading season info')
+
+        for episode in season_info['episodes']:
+            yield self.url_result(f'minitv:{episode["contentId"]}', ie=MiniTVIE.ie_key())
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:285]:
```suggestion
            yield self.url_result(f'minitv:{episode["contentId"]}', MiniTVIE, episode['content_id'])
```

File: yt_dlp/extractor/minitv.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+      seasonId
+    }
+  }
+}
+''',
+            },
+            note='Downloading series info')
+
+        for season in season_info['seasons']:
+            yield self.url_result(f'minitv:season:{season["seasonId"]}', ie=MiniTVSeasonIE.ie_key())
```
Comment (pukkandan) [inline yt_dlp/extractor/minitv.py:333]: same

**Review by pukkandan — CHANGES_REQUESTED** [review]
Also, better to rename extractors to AmazonMiniTV imo

File: yt_dlp/extractor/minitv.py [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+        self._download_webpage(
+            'https://www.amazon.in/minitv',
+            None,
+            headers=self._HEADERS,
+            note='Downloading webpage')
+
+        self.session_id = self._get_cookies('https://www.amazon.in')['session-id'].value
+
+    def _real_extract(self, url):
+        asin = f'amzn1.dv.gti.{self._match_id(url)}'
```
[the remark this answers is not in the document]
Comment (Lesmiscore) [reply yt_dlp/extractor/minitv.py:116]: https://en.wikipedia.org/wiki/Amazon_Standard_Identification_Number

File: yt_dlp/extractor/minitv.py (hunk 6 above)
Comment (nyuszika7h) [reply yt_dlp/extractor/minitv.py:9]: Just for consistency with `minitv:season` and `minitv:series`.

File: yt_dlp/extractor/minitv.py (hunk 7 above)
Comment (nyuszika7h) [reply yt_dlp/extractor/minitv.py:230]: I could add it just in case, but the URLs you get from the website always seem to point to a single episode.

Comment (pukkandan) [reply yt_dlp/extractor/minitv.py]: Write an ie_desc similar to this so users know what prefix to use. (by looking at supportedsites)

```
IE_DESC = 'Amazon MiniTV Series, "minitv:season:" prefix'
```

File: yt_dlp/extractor/minitv.py (hunk 6 above)
Comment (pukkandan) [reply yt_dlp/extractor/minitv.py]: ig it's fine

File: yt_dlp/extractor/amazonminitv.py [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+      seasonId
+    }
+  }
+}
+''',
+            },
+            note='Downloading series info')
+
+        for season in season_info['seasons']:
+            yield self.url_result(f'amazonminitv:{season["seasonId"]}', AmazonMiniTVSeasonIE, season['seasonId'])
```
Comment (GautamMKGarg) [inline yt_dlp/extractor/amazonminitv.py:323]: Series code is not working, changing this line with below will fix it.

`yield self.url_result(f'amazonminitv:season:{season["seasonId"]}', AmazonMiniTVSeasonIE, season['seasonId'])`

Comment (GautamMKGarg) [inline yt_dlp/extractor/amazonminitv.py:323]:
```suggestion
            yield self.url_result(f'amazonminitv:season:{season["seasonId"]}', AmazonMiniTVSeasonIE, season['seasonId'])
```

Series code is not working. This change will fix it

**Review by GautamMKGarg — CHANGES_REQUESTED** [review]
Bugfix while fetching series

**Comment by GautamMKGarg** [comment]
Hi @pukkandan

@nyuszika7h is currently buzy with other things. Only small part of work is left rest is already completed by @nyuszika7h. Could you please complete it?

**Comment by Lesmiscore** [comment]
@GautamMKGarg Could you please check all the rest are working now?

**Comment by pukkandan** [comment]
Anyone is free to make necessary changes and open a new PR

**Comment by GautamMKGarg** [comment]
> @GautamMKGarg Could you please check all the rest are working now?

Yes, Working fine

**Comment by Lesmiscore** [comment]
:facepalm: Looks like the gh command fucked up the repo and I accidentally pushed to this PR

**Comment by pukkandan** [comment]
That's fine. The "make another PR" is for other contributors. You can directly clean up this PR and merge it if u want

**Comment by GautamMKGarg** [comment]
Thanks everyone.

**Pull request merged**